text/000-bundle-field-api.md

```diff
@@ -0,0 +1,159 @@
+- 000 Bundle Field API
+- Start Date: 2020-03-10
+- RFC PR: (leave this empty)
+- Chisel-lang Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes add `Field` as an extension to the `Bundle` construction API to specify
+precisely those parts of a `Bundle` that are actually intended to be functioning fields
+
+# Motivation
+[motivation]: #motivation
+
+The main goal of this addition is to further reduce the number of situations
+where a developer needs to write a `cloneType` method for custom `Bundle`s.
+It is further motivated by the difficulty in recognizing problematic constructions and
+crafting comprehensible error messages when they occur.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+It is proposed that an optional binding operator `Field` be added to the `Bundle` construction API.
+`Field` is similar to the binding operators `IO`, `Input`, `Output`, etc.
+It will allow with better semantics, which will coexist with the old semantics indefinitely. 
+
+Simple example:
+```scala
+class MyBundle extends Bundle {
+  val x = UInt(32.W)
+  val y = UInt(16.W)
+  val flag = Bool()
+}
+```
+would optionally become.
+```scala
+class MyBundle extends Bundle {
+  val x = Field(UInt(32.W))
+  val y = Field(UInt(16.W))
+  val flag = Field(Bool())
+}
+```
+
+Notable in this the following, `gen` and `constant` are not treated as fields of `MyFancierBundle`
+
+```scala
+class MyFancierBundle(val gen: Data) extends Bundle {
+  val x = Field(UInt(32.W))
+  val y = Field(UInt(16.W))
+  val flag = Field(Bool())
+  val constant = 7.U
+}
+```
+
+The new semantics are, tentatively:
+
+`Fields` are all or nothing, if you define any `Bundle` element as a `Field`, you must define all Bundle elements as Fields. The compiler will error out if you either have vals of type Data that aren't declared as a Field, or if you have Fields that weren't autodetected to be Bundle fields (which is a limitation of reflection-based detection).
```

Comment:
Is this analysis possible? If there are hardware elements that are used a generators they will not be wrapped in Field.

Reply:
Yeah i kinda thought the point of this was, in part, to help disambiguate between these two cases, not to forbid having non-field data members altogether. I'm not sure it is sufficient to just support the gen and constant cases (?). Seems to me like we also need a `NotAField`?